# This Jupyter Notebook is about Bayesian Data Analysis for neuroscience data

## Introduction

This notebook is part of a 20-week internship project carried out at Ulster University.
The main goal of the project is to make advanced Bayesian statistical models more accessible
to experimental neuroscientists through user-friendly code, tutorials, and examples.

Specifically, this notebook focuses on applying existing Bayesian models to neuroscience datasets
using libraries in Python.

## Objectives

1. Apply the existing Bayesian models to a neuroscience dataset from scratch,
   documenting each step as if it were for a beginner user.

2. Design a simple and reproducible analysis pipeline using PyMC.

3. Produce clear, well-documented code that can later be integrated into
   an interactive tutorial or a web application.

## Author

- Mathis DA SILVA
- Ulster University Internship (July–December 2025)
- Supervisors: Dr. Cian O'Donnell & Dr. Conor Houghton

## References

- [Dataset from "Classification of psychedelics and psychoactive drugs based on brain-wide imaging of cellular c-Fos expression"](https://www.nature.com/articles/s41467-025-56850-6#Sec25)
- [Hierarchical Bayesian modeling of multi-region brain cell count data](https://elifesciences.org/reviewed-preprints/102391v1)
- [Statistical Rethinking 2023 PDF](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf)
- [Statistical Rethinking 2023 Videos](https://www.youtube.com/watch?v=FdnMWdICdRs&list=PLDcUM9US4XdPz-KxHM4XHt7uUVGWWVSus)

Here, we call libraries that we will use in this notebook for the moment.

In [1]:
import math
import numpy as np
import pymc as pm
import pandas as pd

WARNING (pytensor.configdefaults): g++ not available, if using conda: `conda install gxx`
WARNING (pytensor.configdefaults): g++ not detected!  PyTensor will be unable to compile C-implementations and will default to Python. Performance may be severely degraded. To remove this warning, set PyTensor flags cxx to an empty string.


We will use the dataset from the paper "Classification of psychedelics and psychoactive drugs based on brain-wide imaging of cellular c-Fos expression".



In [2]:
dataset = pd.read_excel('data/dataset_neuroscience_vo.xlsx')

dataset

abbreviation                      region name  brain area 5MEO1 count  \
0            FRP     Frontal pole cerebral cortex      Cortex        9574   
1            ILA                 Infralimbic area      Cortex       12138   
2           ORBl        Orbital area lateral part      Cortex       48129   
3           ORBm         Orbital area medial part      Cortex       17225   
4          ORBvl  Orbital area ventrolateral part      Cortex       32690   
..           ...                              ...         ...         ...   
311         LING                      Lingula (I)  Cerebellum          52   
312           DN                  Dentate nucleus  Cerebellum          51   
313          NOD                      Nodulus (X)  Cerebellum        1528   
314         VeCB      Vestibulocerebellar nucleus  Cerebellum          76   
315          NaN                              NaN         NaN      Female   

    5MEO2 count 5MEO3 count 5MEO4 count 5MEO5 count 5MEO6 count 5MEO7 count  \
0          7781       17598        4425        7428        8302        4288   
1          6742       28070        1685       15612       17191        6061   
2         45849      120147       28655       40438       54206       39938   
3          8551       34163        6330       14908       23250        7993   
4         24460       58132       16015       24182       31926       15148   
..          ...         ...         ...         ...         ...         ...   
311          65         317         221          63         131           2   
312           8         926         144          28          37         185   
313        1153        3218        2399         501         562         111   
314          67         163         111          66         284           0   
315      Female      Female        Male        Male        Male     Unknown   

     ... PSI7 count PSI8 count SAL1 count SAL2 count SAL3 count SAL4 count  \
0    ...       3367       4342       7404       4925      12521      10363   
1    ...       7591       5778       9665       8049      10853       2844   
2    ...       8291      14603      56825      30618      58755      14705   
3    ...       4878       7177      13035      16101      14017       7855   
4    ...       5081       9116      37775      26349      33593      10743   
..   ...        ...        ...        ...        ...        ...        ...   
311  ...        150        112         51          7        119         70   
312  ...         24          8         85         21         19         15   
313  ...       2318       1105        369        185        610         34   
314  ...         16         37         43          2         34         27   
315  ...    Unknown    Unknown     Female     Female     Female       Male   

    SAL5 count SAL6 count SAL7 count SAL8 count  
0         4562      14383        789       6067  
1        15747      15412      11667      21630  
2        26686      59049       5192      36019  
3        14478      30999      10051      28545  
4        21789      37644       9320      23276  
..         ...        ...        ...        ...  
311         22        233        114        255  
312          6          6          0         12  
313       7016       1493       1293       6273  
314          3          3          0         31  
315       Male       Male    Unknown    Unknown  

[316 rows x 67 columns]

#### Indications:

Previously, we added the dataset. In which, the first three columns represent brain regions with name and abbreviation. Others represent mice group by drugs as **MDMA**, **Ketamine**, **Fluoxetine**, ...

There are **64 mice** in total, and each mouse has a value for each brain region. The values represent the number of cells expressing c-Fos, a marker of **neuronal activity**. Plus, there are **315 brain regions** in the dataset.

### Statistical Models
---

We will use a **Hierarchical Bayesian Model** to analyze the dataset. The model will allow us to account for the hierarchical structure of the data, where measurements are nested within brain regions and mice. Models are: **Poisson**, **Horseshoe** and **Zero-Inflated Poisson (ZIP)**.

For the moment, we will use a part of the dataset, specifically the first 20 brain regions and 2 groups of mice.

In [3]:
dataset1 = pd.read_excel('data/dataset_neuroscience_1.xlsx')

dataset1

abbreviation                                region name brain area  \
0           FRP               Frontal pole cerebral cortex     Cortex   
1           ILA                           Infralimbic area     Cortex   
2          ORBl                  Orbital area lateral part     Cortex   
3          ORBm                   Orbital area medial part     Cortex   
4         ORBvl            Orbital area ventrolateral part     Cortex   
5           AId         Agranular insular area dorsal part     Cortex   
6           AIp      Agranular insular area posterior part     Cortex   
7           AIv        Agranular insular area ventral part     Cortex   
8        RSPagl  Retrosplenial area lateral agranular part     Cortex   
9          RSPd             Retrosplenial area dorsal part     Cortex   
10         RSPv            Retrosplenial area ventral part     Cortex   
11           PL                             Prelimbic area     Cortex   
12         VISa                              Anterior area     Cortex   
13          TEa                 Temporal association areas     Cortex   
14         PERI                            Perirhinal area     Cortex   
15          ECT                            Ectorhinal area     Cortex   
16           PA                Posterior amygdalar nucleus     Cortex   
17          BMA               Basomedial amygdalar nucleus     Cortex   
18          BLA              Basolateral amygdalar nucleus     Cortex   
19           LA                  Lateral amygdalar nucleus     Cortex   
20          EPv          Endopiriform nucleus ventral part     Cortex   
21          NaN                                        NaN        NaN   

   A-SSRI1 count A-SSRI2 count A-SSRI3 count A-SSRI4 count A-SSRI5 count  \
0           4297          2320          1873          3262          2331   
1           5028         11432         14109         13773         13181   
2          16644         37202         28689         25857         23043   
3           6143         12661         17336         13613         16076   
4          11682         24502         36990         24050         20598   
5          14605         15198         11975         12488         28418   
6          16409          9414          7116         13154         24693   
7           9821         15879         11689         15838         19235   
8          15133         24749         25622         30641         13331   
9          18329         25369         23375         35179         16741   
10         23408         28252         31993         37309         17451   
11         12922         24065         29145         31021         30130   
12          2548         13042         20386         18868          9565   
13         35667         47646         53008         56115         80049   
14          7133          2429          2469          4255          6862   
15         17644         14098         17772         17611         30319   
16          2922          3707          8779          3319          9481   
17          6143         11110         17309          9851         17434   
18          6710          8568         16714          8613         23019   
19          4952          6527         11122          8496         15260   
20          3448          4398          6609          5477          9002   
21          Male          Male          Male          Male        Female   

   A-SSRI6 count A-SSRI7 count A-SSRI8 count C-SSRI1 count C-SSRI2 count  \
0           1737          6635          1847           775          1302   
1          20988          7812          9091          3352          6365   
2          15582         30958         17778          9138          9322   
3          16278         14271          9314          4652          8659   
4          16528         28842         10982         13188         13533   
5          14536         13091         20905          3899          5779   
6          17012          7253         188

We will prepare the data for the model.

Now, we will build Poisson model using PyMC.

Here, there are some visulization of this model:

\begin{gather*}
y_i \sim Poisson(\lambda_i)\\
log(\lambda_i) = E_i + \gamma_i\\
\theta_{rg} \sim Normal(5, 2)\\
\tau_{rg} \sim HalfNormal(log(1.05))\\
\gamma_i \sim Normal(\theta_{r[i]g[i]}, \tau_{r[i]g[i]})\\
\end{gather*}

In [ ]:
with pm.Model() as poisson_model:



Now, we will build Horseshoe model using PyMC.

Here, there are some visulization of this model:

\begin{gather*}
y_i \sim Poisson(\lambda_i)\\
log(\lambda_i) = E_i + \gamma_i\\
\theta_{rg} \sim Normal(5, 2)\\
\tau_{rg} \sim HalfNormal(log(1.05))\\
\kappa_i \sim HalfNormal(1)\\
\gamma_i \sim Normal(\theta_{r[i]g[i]}, \kappa_i\times\tau_{r[i]g[i]})\\
\end{gather*}

In [ ]:
with pm.Model() as horseshoe_model:


Now, we will build Zero-inflated Poisson model using PyMC.

Here, there are some visulization of this model:

\begin{gather*}
y_i \sim ZIPoisson(\lambda_i,\pi)\\
log(\lambda_i) = E_i + \gamma_i\\
\pi \sim Beta(1,5)\\
\theta_{rg} \sim Normal(5, 2)\\
\tau_{rg} \sim HalfNormal(log(1.05))\\
\gamma_i \sim Normal(\theta_{r[i]g[i]}, \tau_{r[i]g[i]})\\
\end{gather*}

In [ ]:
with pm.Model as zeroinflatedpoisson_model:
